enhancements/olm-toggle-copied-csvs.md

```diff
@@ -0,0 +1,204 @@
+
+#### Operator Cardinality
+
+Today, Operator resources are generated on a per-Subscription basis. Since CSVs -- and thus Copied CSVs -- can exist without matching Subscriptions, OLM's controllers will be extended to generate additional Operator resources for these "free CSVs" such that:
+
+- there exists at least one `Operator` resource for every CSV
+- there exists at least one `Operator` resource for every `Subscription`
+
+That is, all CSVs installed by a given `Subscription` share an `Operator` resource __and__ any CSVs that weren't installed by a `Subscription` are related to their own `Operator` resource.
+
+#### Orphaned CSVs
+
+Orphaned CSVs are CSVs that were installed via a `Subscription` that no longer exists. Since the lifetimes -- i.e. the period of existence on a cluster --  of `Operator` resources aren't directly tied to the lifetimes of the resources that triggered their generation, the `Operator` resource(s) associated with Orphaned CSVs will continue to exist _after_ a `Subscription` is deleted. If that `Operator` resource is deleted explicitly, OLM will generate another as per the rules [outlined above](#operator-cardinality).
+
+#### Copied RBAC 
+
+In addition to operator discoverability, Copied CSVs also serve as convienient vehicles for the garbage collection of Copied RBAC. These are resources that are copied along with the CSV to provide the operator access to a namespace. To support the loss of Copied CSVs, the RBAC copying logic in OLM will be extended to write additional owner references from copied resources to the related Operator resource. 
+
+#### The Toggle
+
+A feature toggle will be added to control OLM's use of Copied CSVs. When enabled (toggled on), OLM will delete all existing Copied CSVs and will not generate any more. When disabled (toggled off), OLM will generate Copied CSVs (much like it does today).
+
+When toggled (on then off), OLM will recreate any missing Copied CSVs and reassociate the owner references of related copied resources with them (i.e. the inverse operation of the one described in the [Copied RBAC](#copied-rbac) section).
+
+The toggle will be disabled by default.
+
+The toggle can be set from a command-line option on the olm-operator binary.
+
+#### Config CRD w/ Toggle
+
+A novel cluster-scoped `OLMConfig` CRD will be added to OLM. The resource type it defines will allow users to apply a limited set of configurations to the OLM instance on their cluster by creating or modifying a singlton (of that type) on the cluster.
+
+The name of that singleton will be `cluster` and OLM will ignore all other `OLMConfig` resources on a cluster.
+
+Its spec will contain a Copied CSV feature toggle field.
+
+e.g.
+
+```yaml
+apiVersion: operators.coreos.com/v1
+kind: OLMConfig
+metadata:
+  name: cluster
+spec:
+  features:
+    'disable-copied-csvs': true
+```
+
+When both the [command-line option](#the-toggle) _and_ `cluster` resource toggle field are present, the latter takes precedence.
```

Comment:
Not necessarily opposed, but what's the motivation for having two ways to set the toggle? Seems like it would lead to user confusion due to precedence ambiguity (i.e. does anyone actually read the docs?! 😛)

Reply:
This is a fair point.

My thinking at the time I wrote this:

- lets you set config atomically; i.e. if I don't want to impose ordering rules between OLM's deployment and the config api
- you could compose a config controller that updates an independent controller, rather than watching directly in OLM
  - this also gives you the option of making all command-line options opaque in the config schema (for better or worse)
- feature toggles are typically implemented as command line options (as far as I'm aware)

Granted, that's all pretty watery. I'd be inclined to remove it if you feel it will negatively impact UX.

Reply:
Having started implementing this feature with both the flag and the CRD, I support removing the flag. The existing implementation leads to a non-intuitive user experience.